File: solutions/4133-minimum-generations-to-target-point/minimum-generations-to-target-point.py
class Solution:
    def cal(self, a, b):
        xa, ya, za = a // 100, (a % 100) // 10, a % 10
        xb, yb, zb = b // 100, (b % 100) // 10, b % 10

        return (100 * ((xa + xb) // 2) + 10 * ((ya + yb) // 2) + (za + zb) // 2)

    def minGenerations(self, points, target):
        s = set()

        for x, y, z in points:
            s.add(100 * x + 10 * y + z)

        t = 100 * target[0] + 10 * target[1] + target[2]

        ans = 0

        while t not in s:
            temp = set(s)

            arr = list(s)

            for i in range(len(arr)):
                for j in range(i + 1, len(arr)):
                    temp.add(self.cal(arr[i], arr[j]))

            if temp == s:
                return -1

            s = temp
            ans += 1

        return ans

File: solutions/4133-minimum-generations-to-target-point/minimum-generations-to-target-point.java
class Solution {

    private int cal(int a, int b) {
        int xa = a / 100;
        int ya = (a % 100) / 10;
        int za = a % 10;

        int xb = b / 100;
        int yb = (b % 100) / 10;
        int zb = b % 10;

        return 100 * ((xa + xb) / 2) + 10 * ((ya + yb) / 2) + (za + zb) / 2;
    }

    public int minGenerations(int[][] points, int[] target) {

        Set<Integer> s = new HashSet<>();

        for (int[] p : points) s.add(100 * p[0] + 10 * p[1] + p[2]);

        int t = 100 * target[0] + 10 * target[1] + target[2];

        int ans = 0;

        while (!s.contains(t)) {

            Set<Integer> temp = new HashSet<>(s);

            List<Integer> arr = new ArrayList<>(s);

            for (int i = 0; i < arr.size(); i++) {
                for (int j = i + 1; j < arr.size(); j++) {
                    temp.add(cal(arr.get(i), arr.get(j)));
                }
            }

            if (temp.equals(s)) {
                return -1;
            }

            s = temp;
            ans++;
        }

        return ans;
    }
}